padinfo/boardgenerator.py

```diff
@@ -0,0 +1,69 @@
+class BoardGenerator(object):
+    """Generate a Dawnglare link with optional inversion."""
+
+    allowed_letters = "RGBLDHXJMPZ "
+    inversion = "-1"
+    dawnglare_link = "https://pad.dawnglare.com/?patt={}&showfill=1"
+
+    def __init__(self, board):
+        self.board = board.split()
+        self.board_height = None
+        self.board_width = None
+        self.invalid_board = False
+        self.invert = BoardGenerator.inversion in self.board
+        self.clean_board = self.get_clean_board()
+        self.invalid_orbs = not all(letter in BoardGenerator.allowed_letters for letter in self.clean_board)
+        self.board_length = len(self.clean_board)
+        self.eval_board_size()
+        self.link = self.get_link()
+
+    def get_clean_board(self):
+        board = self.board
+
+        if self.invert:
```

Comment:
EXTREMELY OPTIONAL!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

okay so, what you have here is TOTALLY FINE and works. But, there's a small "separation of concerns" that I don't love, and you can either do it now, or, you could come back in a couple projects from now and redo it later. If it doesn't 100% make sense now I'd rather you do it when it does make sense instead of doing it just because I'm saying so.

* In the `init` you get `self.invert` as whether `-1` is in the string (btw, I LOVE that you set that as a class constant instead of hard-coding the string `-1` in the lookup!!). That on its own is fine, but
* Then here, in a totally different location, we're dealing with the consequences of that inversion being there. Now, this is the ONLY place where we're dealing with those consequences.
	* If there were multiple places where we cared about the fact that the inversion were present, it might make sense to do a dedicated class constant and use it in multiple places
	* Also we could potentially set a class constant for e.g. debugging purposes, so it may make sense to set the class constant to None in the init and then set the correct value here
	* The problem is, we've detected the presence of the -1 way far away in the init, and then we're dealing with it here, plus modifying the string here, and so these two bits of logic are really separate from each other. If we were to change how setting inversion were done (say in the future it was done by wrapping the thing in `()`) then the pieces of logic are super far away from each other
* So, I would like to see a refactor that brought those pieces of logic closer to each other. Again the class constant isn't necessarily a bad idea if you wanna debug, but I'd detect inversion presence near the rest of the inversion logic (and then set the value of the constant there)

Again there is absolutely no need to resolve this issue right now, and if you don't like 100% totally understand it, please postpone it until you do!!

Reply:
To be clear, do you mean this line in get_clean_board? `self.board.remove(self.inversion)`

I'm already thinking through some refactors based on your comments on being able to take an input not separated with spaces. If that line is the problem, then I think I know how to resolve this, but I want to be sure first.

Reply:
So the problem isn't any one thing in isolation, but that

```
self.invert = self.inversion in self.board```

is very far away from

```
        if not self.invert:
            return ''.join(self.board)

        self.board.remove(self.inversion)
```

AND ALSO

that `self.invert` is never used except in conjunction with this other block of code.

So separating the assignment of `self.invert` (and therefore the logic to determine its value) so very far away from where we're using it is kinda sus, because there's no real reason to.

If we were using `self.invert` all over the place, then by necessity these things might be far away from each other, but, that's not the case here, so we should do them close to each other.

Reply:
Take a look at my latest commit and see if it resolves your concern. I did a bit bigger refactor with the inversion logic to handle not having spaces in the input. I'm stripping the inversion characters right away after I check for its presence so that I can do the rest of the necessary calculations on a clean string.

Reply:
yep, this looks a lot better now